Support localized fields in API responses

bobsilverberg opened the pull request:
Fixes #6422

willdurand commented on src/core/types/addons.js:
```diff
@@ -54,6 +55,24 @@ type EventDataType = {|
+  thumbnail_src: string,
+  thumbnail_w: number,
+  title: string,
+  w: number,
```
If this is our internal type, can we have a better prop name than a single char?

bobsilverberg replied:
Yes, but before I change it I want to confirm whether you are okay with keeping these changes in this patch, or if we should leave them to a follow-up patch.

bobsilverberg replied:
Ah, I believe we're using `w` and `h` here because those are the property names that the `PhotoSwipeGallery` component expects, so we either need to keep them that way, or introduce another transformer to convert from `PreviewType` to an object that works with `PhotoSwipeGallery`. The latter seems totally unneccesary. How about I add a comment to the type to explain that the prop names are those needed for the `PhotoSwipeGallery` component?

willdurand replied:
You're right. I forgot about that (although I thought about it after having left the comment). Yes, a comment sounds good, thanks

willdurand commented on src/core/types/addons.js:
```diff
@@ -54,6 +55,24 @@ type EventDataType = {|
+|};
+
+export type PreviewType = {|
+  h: number,
```
If this is our internal type, can we have a better prop name than a single char?

bobsilverberg replied:
Yes, but before I change it I want to confirm whether you are okay with keeping these changes in this patch, or if we should leave them to a follow-up patch.

willdurand commented on src/core/types/addons.js:
```diff
@@ -122,6 +141,15 @@ export type PartialExternalAddonType = {|
  */
 export type AddonType = {|
   ...ExternalAddonType,
+  // normalized l10n fields
```
This is great but not true in the current patch. The assumption was that `selectLocalizedContent()` return type would be `string | null`. I think it's what we want and I explained why in a previous comment (so let's keep these changes here and update the selector a bit)

willdurand commented on tests/unit/amo/components/TestAddAddonToCollection.js:
```diff
@@ -38,7 +38,7 @@ describe(__filename, () => {
   const getProps = (customProps = {}) => {
     return {
-      addon: createInternalAddon(fakeAddon),
+      addon: createInternalAddonWithLang(fakeAddon),
```
Awesome. I was thinking about doing something like that too.

Nit: I'd have used a slightly different name like `createInternalAddonWithDefaultLang` because, to me, `WithLang` means we'd pass a lang to the function.

bobsilverberg replied:
As you saw later, this function does accept a `lang`, it's just that most of the time we don't need to specify one so we don't pass it and it uses a default. Can I mark this comment as resolved?

willdurand replied:
> Can I mark this comment as resolved?

yes

willdurand commented on tests/unit/amo/components/TestAddonMoreInfo.js:
```diff
@@ -139,17 +140,17 @@ describe(__filename, () => {
+    const addon = createInternalAddonWithLang({
       ...fakeAddon,
-      support_url: 'http://support.hampsterdance.com/',
+      support_url: createLocalizedString('http://support.hampsterdance.com/'),
```
I haven't seen this function yet. I suppose it uses a default lang? How do we sync `createInternalAddonWithLang` and `createLocalizedString`? Maybe we could name these functions `createInternalAddonWithDefaultLang` and `createLocalizedStringWithDefaultLang`?

It's verbose but explicit. Other ideas welcomed, though.

bobsilverberg replied:
As discussed above (and I think as you noticed later), `createLocalizedString` does optionally accept a `lang` if we need to synchronize. Most of the time we don't, which makes the test code and its usage shorter. To me it's fine for a test to use both `createInternalAddonWithLang` and `createLocalizedString` without passing a `lang` to either, and assume that they are synchronized behind the scenes (which they are). Do you agree?

willdurand replied:
I agree.

willdurand commented on tests/unit/amo/components/TestCollectionDetails.js:
```diff
@@ -10,11 +10,11 @@ import {
 import {
   createFakeCollectionDetail,
   createFakeEvent,
+  createInternalCollectionWithLang,
```
`WithDefaultLang`? 😋

bobsilverberg replied:
See above.

willdurand commented on tests/unit/amo/components/TestInstallButtonWrapper.js:
```diff
@@ -79,14 +80,14 @@ describe(__filename, () => {
+      addon: createInternalAddonWithLang(addon),
       clientApp,
-      currentVersion: createInternalVersion(addon.current_version),
+      currentVersion: createInternalVersionWithLang(addon.current_version),
```
`WithDefaultLang`?

bobsilverberg replied:
See above.

willdurand commented on tests/unit/amo/components/TestSponsoredAddonsShelf.js:
```diff
@@ -64,6 +65,7 @@ describe(__filename, () => {
   };
 
   const _loadPromotedExtensions = ({ addons = [] }) => {
+    store.dispatch(setLang('en-US'));
```
`dispatchClientMetadata()` (which is called in the test setup method) already dispatches this action. Why is this needed here?

bobsilverberg replied:
You're right. I added that prematurely. I'll remove it.

willdurand commented on tests/unit/amo/pages/TestHome.js:
```diff
@@ -649,14 +649,14 @@ describe(__filename, () => {
       expect(heroRecommendation).toHaveProp(
         'shelfData',
-        createInternalHeroShelves(heroShelves).primary,
+        createInternalHeroShelves(heroShelves, 'en-US').primary,
```
Should we have a `createInternalHeroShelvesWithDefaultLang` helper?

bobsilverberg replied:
Yes, I'll add one.

willdurand commented on tests/unit/amo/pages/TestLandingPage.js:
```diff
@@ -53,6 +55,7 @@ describe(__filename, () => {
     errorHandler = createStubErrorHandler(),
     ...otherParams
   } = {}) => {
+    store.dispatch(setLang('en-US'));
```
Why is this needed?

bobsilverberg replied:
Hmm, I guess this is another case of me adding this prematurely. I thought I needed it to get tests to pass, but apparently not. I'll remove it.

willdurand commented on tests/unit/helpers.js:
```diff
@@ -1354,3 +1369,23 @@ export const fakeSponsoredShelf = Object.freeze({
   impression_url: 'https://mozilla.org/',
 });
+
+export const createInternalAddonWithLang = (addon, lang = 'en-US') => {
```
Oh,I see why it is not suffixed with `WithDefaultLang`. I guess that's fine then?

willdurand commented on tests/unit/helpers.js:
```diff
@@ -37,23 +38,29 @@ import {
   'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.1';
 export const sampleUserAgentParsed = UAParser(sampleUserAgent);
 
+export const createLocalizedString = (string, lang = 'en-US') => {
```
Instead of having `'en-US'` defined in different functions, maybe we could introduce a constant (`DEFAULT_LANG` or `DEFAULT_LANG_IN_TESTS`) in this file? We could use `'en-US'` or `config.get('defaultLang')` as a value.

bobsilverberg replied:
Yeah, I like this idea. I'll change it.

willdurand reviewed (commented):
Awesome, thanks!

I have some comments but this is looking very good.

bobsilverberg commented:
> Awesome, thanks!
> 
> I have some comments but this is looking very good.

Thanks for the initial look @willdurand. I'll respond to your comments. There is still work to be done to add coverage for a bunch of things. What I've done so far is to just change all the code and get all the tests passing, but there is definitely missing coverage in the current patch.

bobsilverberg commented on src/amo/reducers/addonsByAuthors.js:
```diff
@@ -21,6 +23,7 @@ export type AddonsByAuthorsState = {|
 |};
 
 export const initialState: AddonsByAuthorsState = {
+  lang: '',
```
Done in the next commit.

bobsilverberg commented on src/amo/components/ScreenShots/index.js:
```diff
@@ -59,36 +60,7 @@ export const PHOTO_SWIPE_OPTIONS = {
-  }));
-
-export const thumbnailContent = (item: Preview): React.Node => (
+export const thumbnailContent = (item: PreviewType): React.Node => (
```
Yes and no. `caption` is a localized field, so yes I do need to touch this code. While I was editing it I realized that it's not optimal and not the way we do it anywhere else, so I decided to fix that with a refactor. Considering how big this patch is already, and the fact that pretty much everything in the site will need to be tested, I'm not sure there's a need to extract these changes out into a separate patch. If you feel really strongly about it I can do it, but I feel like this patch is a good opportunity to clean some things up (like this) as long as we're touching the files.

willdurand replied:
Although I think it'd be better to have separate patches, let's keep everything as it is currently (i.e. no other patch, we keep all changes/cleanups here)

bobsilverberg commented on src/core/imageUtils.js:
```diff
@@ -25,8 +25,8 @@ export const getPreviewImage = (
-      // The preview.image_size[0] is the image width.
-      addon.previews.findIndex((preview) => preview.image_size[0] === width);
+      // The preview.w is the image width.
+      addon.previews.findIndex((preview) => preview.w === width);
```
Yes, as mentioned above, `caption` is a localized string.

bobsilverberg commented on src/core/reducers/utils.js:
```diff
@@ -0,0 +1,19 @@
+) => {
+  invariant(lang, 'lang must not be empty');
+  if (!field) {
+    return undefined;
```
I did this to deal with `notes` in CollectionAddon not allowing `null`, but going through the types again I think it actually should allow `null`, so I think I can change this back.

willdurand commented:
> Thanks for the initial look @willdurand. I'll respond to your comments. There is still work to be done to add coverage for a bunch of things. What I've done so far is to just change all the code and get all the tests passing, but there is definitely missing coverage in the current patch.

Yep, it's going pretty well so far. Adding new test coverage is easy compared to refactoring an exiting test suite :) Good job!

bobsilverberg commented:
> Yep, it's going pretty well so far. Adding new test coverage is easy compared to refactoring an exiting test suite :) Good job!

Thanks @willdurand! I have addressed all of your issues, including adding a comment for https://github.com/mozilla/addons-frontend/pull/9909#discussion_r539126081 (which I still need you to confirm is okay). I'll move forward with adding more coverage.

bobsilverberg commented:
This is ready for a final review.

willdurand reviewed (commented):
The language tools page does not work for me.

bobsilverberg commented:
@willdurand I fixed the problem with Language Tools, and also a problem I discovered with Search box autocomplete. I believe I have tested everything locally now, so this is good for another look.

willdurand reviewed (approved):
looks good, thanks

bobsilverberg commented:
Thanks @willdurand. I guess we land it on dev and see how it goes there.
